### PR TITLE
fix: improve perf loading

### PR DESCRIPTION
**Pull request by guilhem-barthes**
## Companion PR

- https://github.com/Substra/substra-frontend/pull/180

## Description

Try to improve loading in performance page. Two strategies have been put in place:

1. Using prefetch for FunctionOutput and using it in the serializer (instead of making a request for each function on FunctionOutput)
2. Removing `data_samples` from the request, as it doesn't seem to be used in the frontend.

The point 2 is more controversial, as it breaks the current architecture of returning linked data on eac hrequest. However, it implies reaching a n+ 3 level (as the data_samples are ordered on `api_taskdatasample.order`, which is maybe not needed .

## Checklist

- [x] [changelog](../CHANGELOG.md) was updated with notable changes
- [ ] documentation was updated

### CODE REVIEW

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

**Comment by guilhem-barthes** [comment]
/e2e --tests sdk,frontend,substrafl --benchmarks mnist,camelyon

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

File: backend/api/tests/views/test_views_performance.py [hunk 1]
```diff
@@ -347,3 +350,64 @@ def test_performance_export_filter_and(self):
         content_list = list(response.streaming_content)
         self.assertEqual(len(content_list), 4)
         self.assertTrue(status in str(content_list[1]))
+
+
+@pytest.fixture
```
Comment (SdgJlbl) [inline backend/api/tests/views/test_views_performance.py:355]: If we use this fixture in multiple places, maybe it's worth putting it in `conftest.py`?

File: backend/api/tests/views/test_views_performance.py [hunk 2]
```diff
@@ -347,3 +350,64 @@ def test_performance_export_filter_and(self):
...
+
+@pytest.mark.django_db
+def test_n_plus_one_queries_performance_list_empty(authenticated_client, create_compute_plan):
+    # This expects 12 requests even without any performance and 14 if performances are made
+    compute_plan = create_compute_plan(n_task=4)
+    url = reverse("api:compute_plan_perf-list", args=[compute_plan.key])
+    with utils.CaptureQueriesContext(connection) as query:
+        print(authenticated_client.get(url))
+    query_tasks = len(query.captured_queries)
+    assert query_tasks == 12
```
Comment (SdgJlbl) [inline backend/api/tests/views/test_views_performance.py:400]: I've been told by my local DSF fellow that I should not check for an exact number of queries in my tests, since it's typically the kind of things that can change slightly due to unrelated changes somewhere else in the API. Rather, check for a range to ensure that it's not exploding / evolving linearly with the number of tasks / samples.

Comment (SdgJlbl) [reply backend/api/tests/views/test_views_performance.py:400]: Up this comment

File: backend/api/tests/views/test_views_performance.py [hunk 3]
```diff
@@ -350,3 +350,64 @@ def test_performance_export_filter_and(self):
...
+    def _create_compute_task(compute_plan, n_data_sample=4):
+        data_manager = factory.create_datamanager()
+        data_samples = [factory.create_datasample([data_manager]) for _ in range(n_data_sample)]
+        input_keys = {
+            "opener": [data_manager.key],
+            "datasamples": [data_sample.key for data_sample in data_samples],
+        }
+        function = factory.create_function(
+            inputs=factory.build_function_inputs(["datasamples", "opener", "model"]),
+            outputs=factory.build_function_outputs(["performance"]),
```
Comment (ThibaultFy) [inline backend/api/tests/views/test_views_performance.py:366]: test with a dummy output in addition to performance

**Review by ThibaultFy — APPROVED** [review]
lgtm !

File: backend/api/tests/views/test_views_performance.py (hunk 2 above)
Comment (guilhem-barthes) [reply backend/api/tests/views/test_views_performance.py:400]: I like this approach, it also really tests what we want to do - that performances do not impact a lot these requests. I also expect the other PR to decrease this number, going from 12 for an empty compute plan to less

Comment (guilhem-barthes) [reply backend/api/tests/views/test_views_performance.py]: @SdgJlbl , I went to the smallest comparison (e.g. won't fail if the number of calls decrease), but then the difference between with and without performance is only < 3, does it look ok?

Comment (SdgJlbl) [reply backend/api/tests/views/test_views_performance.py]: well, if we did not forget any prefetch_related and with the right model, querying all performance should be in O(1), so possibly only 1 more call compared to not querying it. So yeah, not so easy to test, given that Django is caching at least 4 queries between the first time and the subsequent times.

**Comment by guilhem-barthes** [comment]
/e2e --tests sdk,frontend,substrafl --benchmarks mnist,camelyon

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**
